src/main/java/org/folio/circulation/domain/OverduePeriodCalculatorService.java

```diff
@@ -0,0 +1,108 @@
+package org.folio.circulation.domain;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.folio.circulation.support.Result;
+import org.joda.time.DateTime;
+import org.joda.time.LocalTime;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.Result.succeeded;
+import static org.folio.circulation.support.ResultBinding.flatMapResult;
+import static org.joda.time.DateTimeConstants.MINUTES_PER_DAY;
+import static org.joda.time.DateTimeConstants.MINUTES_PER_HOUR;
+import static org.joda.time.Minutes.minutesBetween;
+
+public class OverduePeriodCalculatorService {
+  private static final int ZERO_MINUTES = 0;
+
+  private final CalendarRepository calendarRepository;
+
+  public OverduePeriodCalculatorService(CalendarRepository calendarRepository) {
+    this.calendarRepository = calendarRepository;
+  }
+
+  public CompletableFuture<Result<Integer>> getMinutes(Loan loan, DateTime systemTime) {
+    if (preconditionsAreNotMet(loan, systemTime)) {
+      return completedFuture(succeeded(ZERO_MINUTES));
+    }
+
+    return getOverdueMinutes(loan, systemTime)
+      .thenApply(flatMapResult(om -> adjustOverdueWithGracePeriod(loan, om)));
+  }
+
+  boolean preconditionsAreNotMet(Loan loan, DateTime systemTime) {
+    DateTime dueDate = loan.getDueDate();
+
+    return dueDate == null
+      || dueDate.isAfter(systemTime)
+      || loan.getOverdueFinePolicy().getCountClosed() == null;
+  }
```

Comment:
Personally, I might rephrase this as the positive statements around the preconditions
* The loan is overdue
* There is a valid overdue fine policy (is that what `loan.getOverdueFinePolicy().getCountClosed() != null` means?

Reply:
An overdue fine policy is still valid if `countClosed` is null, but then we do not know which strategy we should use to calculate overdue time, therefore should return zero immediately.

> Does the loan also need to be closed for the fine to be applied?

As far as I know, overdue fine is calculated in two cases:
- upon check-in
- upon renewal

I case of chech-in, yes, the loan is closed first and then the fine is calculated. I'm not sure about renewal though. I assume the loan is updated, not closed.

Reply:
@OleksandrVidinieiev Ok, thanks for the explanation.

Is the current code intended to accommodate both of those scenarios? If so, how does this code differentiate between these situations?

Reply:
> An overdue fine policy is still valid if `countClosed` is null, but then we do not know which strategy we should use to calculate overdue time, therefore should return zero immediately.

How is the policy valid if it cannot be applied? Maybe I'm using `valid` in a different sense, maybe `applicable` would be a better word.

What does the ` countClosed` property mean? What values can it have?

Reply:
> Is the current code intended to accommodate both of those scenarios?

Yes.

> If so, how does this code differentiate between these situations?

It does not, because it shouldn't. The logic for calculation of overdue period is the same for both scenarios.

Reply:
> > If so, how does this code differentiate between these situations?
> 
> It does not, because it shouldn't. The logic for calculation of overdue period is the same for both scenarios.

That suggests that when a loan is renewed and then returned, the borrower would have been issued two fines (once upon renewal and once upon return). 

Is that the expected behaviour?

Maybe this is a question for later work, rather than now.

Reply:
I believe that renewal process updates loan's due date, so the borrower won't be fined again upon return (if he returns the item before the updated "new" due date).

Reply:
@alexanderkurash Yes, I think you are correct, thank you.